### PR TITLE
added clear-route-status.sh script to hack/

### PR DESCRIPTION
**Pull request by JacobTanenbaum**
Added a script to clear the route status of unwanted router information.

./clear-route-status.sh [namespace] [route-name | ALL]

To clear the status of all routes in all namespaces:
oc get namespaces | awk '{if (NR!=1) print \$1}' | xargs -n 1 -I %% ./clear-route-status.sh %% ALL

To clear the status of all routes in namespace default:
./clear-route-status.sh defalt ALL

To clear the status of route example in namespace default:
./clear-route-status.sh default example

NOTE: if a router that admits a route is running it will immediatly update the cleared route status

### CODE REVIEW

**Comment by JacobTanenbaum** [comment]
@knobunc @smarterclayton

**Comment by JacobTanenbaum** [comment]
@openshift/networking

**Comment by marun** [comment]
cc: @stevekuznetsov

File: hack/clear-route-status.sh [hunk 1]
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
+
+clear_status ()
```
Comment (pecameron) [inline hack/clear-route-status.sh]: I think the form is:
function clear_status () {
}

File: hack/clear-route-status.sh [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+fi
+
+oc proxy > /dev/null &
+PROXY_PID=$!
+
+
+
+if [ "$2" == "ALL" ]; then
+	ROUTES=$(oc get routes --namespace=$1 2> /dev/null | awk '{if (NR!=1) print $1}')
+	if [ -n "$ROUTES" ]; then
```
Comment (pecameron) [inline hack/clear-route-status.sh]: They like the form ${ROUTES}

File: hack/clear-route-status.sh [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+		for ROUTE in $ROUTES; do
```
Comment (pecameron) [inline hack/clear-route-status.sh]: ${ROUTES}

File: hack/clear-route-status.sh [hunk 4]
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
+
+clear_status ()
+{
+	MY_JSON_BLOB=$(curl -s -X GET http://localhost:8001/oapi/v1/namespaces/$1/routes/$2/)
```
Comment (stevekuznetsov) [inline hack/clear-route-status.sh]: What is this hard-coded API server hostname useful for?

Comment (stevekuznetsov) [inline hack/clear-route-status.sh]: Why aren't we using an actual client for this? If we NEED raw, at the least `oc get --raw`

File: hack/clear-route-status.sh [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	if [ $(echo "$MY_JSON_BLOB" |jq '.kind') == '"Route"' ]; then
+		MODIFIED_JSON=$(echo "$MY_JSON_BLOB" | jq 'del(.status.ingress)')
```
Comment (stevekuznetsov) [inline hack/clear-route-status.sh]: Looks like you want to do a `PATCH`

File: hack/clear-route-status.sh [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+		curl -s -X PUT http://localhost:8001/oapi/v1/namespaces/$1/routes/$2/status --data-binary "$MODIFIED_JSON" -H "Content-Type: application/json" > /dev/null
+		echo "route status for route $2 in namespace $1 cleared"
+	else
+		echo "Route $2 not found in namespace $1"
+	fi
+}
+
+USAGE="Usage:
+./clear-route-status.sh [namespace] [route-name | ALL]
```
Comment (knobunc) [inline hack/clear-route-status.sh]: Ah, I think you missed a slight nuance.  It's the _router_ name that we want to take here.

So if I deploy a router named 'bob' and then delete that router, I need to find all routes who have a status from router 'bob' and delete them.

File: hack/clear-route-status.sh (hunk 2 above)
Comment (stevekuznetsov) [inline hack/clear-route-status.sh]: Also always prefer `[[` over `[` unless you need POSIX compliance.

**Comment by stevekuznetsov** [comment]
Shouldn't we be making it so that deleting a router cleans this up, in a controller or something? Is that also being done? What is the lifetime of this sort of script in this repo?

File: hack/clear-route-status.sh [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+	exit
+fi
+
+oc proxy > /dev/null &
+PROXY_PID=$!
+
+
+
+if [ "$2" == "ALL" ]; then
+	ROUTES=$(oc get routes --namespace=$1 2> /dev/null | awk '{if (NR!=1) print $1}')
```
Comment (ramr) [inline hack/clear-route-status.sh]: You can use a template instead of the pipe to awk to get just the names of the routes. 
Example: `oc get route -n default --template="{{range .items}} {{.metadata.name}} {{end}}"`

Comment (stevekuznetsov) [inline hack/clear-route-status.sh]: Since this is a common use-case, you can simply do:

```sh
$ oc get routes -n default -o name
```

Comment (ramr) [inline hack/clear-route-status.sh]: +1 Even better

**Comment by smarterclayton** [comment]
Put this into the images/router or contrib directory, it really doesn't
belong in hack/ which is for build tools

On Dec 13, 2016, at 3:32 PM, Ram Ranganathan <notifications@github.com>
wrote:

*@ramr* commented on this pull request.
------------------------------

In hack/clear-route-status.sh
<https://github.com/openshift/origin/pull/12180>:

> +NOTE: if a router that admits a route is running it will immediatly update the cleared route status
+"
+
+
+if [ "$#" -ne 2 ] || [[ "$@" == *" help "* ]]; then
+	printf "$USAGE"
+	exit
+fi
+
+oc proxy > /dev/null &
+PROXY_PID=$!
+
+
+
+if [ "$2" == "ALL" ]; then
+	ROUTES=$(oc get routes --namespace=$1 2> /dev/null | awk '{if
(NR!=1) print $1}')

+1 Even better

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/12180>, or mute the thread

**Comment by stevekuznetsov** [comment]
After getting a little fill-in as to the background here ... @smarterclayton I don't think asking someone to patch together a script using `curl` and `jq` is really a great user story for "look! you can script our client to do complicated things!". I don't _really_ care what gets put into this script if it's useful for devs, but I feel like at the least this should wait until it can be completed with `oc` and effort should be put into supporting `patch` for sub-resources, etc...

**Comment by smarterclayton** [comment]
We're not going to put this into oc in the near term, this is education for
end users scripting our API.  We will not implement everything in oc, we
will enable our administrators to solve problems.  That will involve
scripting.

Opening user stories to track the things that should be simpler but aren't
is a great step.  It doesn't block this.

**Comment by stevekuznetsov** [comment]
> We're not going to put this into oc in the near term

Do you mean we won't put router status pruning into `oc` or do you mean we won't get support for sub-resources in `oc patch`?

**Comment by smarterclayton** [comment]
Either.

**Comment by stevekuznetsov** [comment]
Given that support in `oc patch` this script could look like:

```sh

for route in $( oc get routes --namespace="${NAMESPACE}" -o name ); do
    oc patch route "${route}" --namespace="${NAMESPACE}" -p "{"status":{"ingress":{}}}"
done
```

**Comment by smarterclayton** [comment]
It doesn't block this PR.  I agree that we should support sub resources in
the CLI.  Independent problems.

**Comment by stevekuznetsov** [comment]
>  I agree that we should support sub resources in the CLI. Independent problems.

OK

**Comment by stevekuznetsov** [comment]
The only frustrating part of this is we _just_ ripped out all of the Bash logic to use certs etc for command-line use of `curl` against the APIServer and declared that from here on out we wouldn't support that in our scripts, opting to use `oc get --raw` everywhere else. We haven't had the need to issue other verbs "raw" yet ...

So a "complete" implementation of this script (if it needs to exist) would have to go back, get all that logic back, and have it inside. There's special logic for MacOS hosts since they require p12 format .... it's a PITA. If this something you envision devs/users using (and not something that sits in an image and uses a serviceaccount token to authenticate) I would suggest looking at `hack/util.sh` maybe 30 commits ago, look for `set_curl_args` function.

**Comment by stevekuznetsov** [comment]
Depending on the requirements of this script and the difficulties with `oc patch`, it may be less work to go upstream and work to get `patch` to support sub-resources and then make your script simple rather than expending a lot of work getting lots of Bash logic built in for using `curl` in this script, then having to revisit it and rip it all out in the future as well.

File: hack/clear-route-status.sh [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+			clear_status $1 $ROUTE
+		done
+	else
+		echo "No routes found for namespace $1"
+	fi
+else
+	clear_status $1 $2
+fi
+
+kill $PROXY_PID
```
Comment (stevekuznetsov) [inline hack/clear-route-status.sh]: You'll want to trap a cleanup function on `EXIT` to make sure this happens.

File: images/router/clear-route-status.sh [hunk 9]
```diff
@@ -0,0 +1,60 @@
+#!/bin/bash
+
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Always better to run in "safe" mode:

```sh
set -o errexit
set -o nounset
set -o pipefail
```

File: images/router/clear-route-status.sh [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+# This allows for the clearing of route statuses, routers don't clear the routes status so some may be stale.
+# Upon deletion of the routes status active routers will immediatly update with a vaild status
+
+clear_status() {
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: We prefer

```sh
function clear_status() {
```

File: images/router/clear-route-status.sh [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    MY_JSON_BLOB=$(curl -s -X GET http://localhost:8001/oapi/v1/namespaces/$1/routes/$2/)
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Does `oc get --raw` not jive with `oc proxy`?

File: images/router/clear-route-status.sh [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    if [[ $(echo ${MY_JSON_BLOB} |jq '.kind') == '"Route"' ]]; then
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: You're hitting the API for a route -- when would this ever not be the case?

File: images/router/clear-route-status.sh [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+./clear-route-status.sh defalt ALL
+
+To clear the status of route example in namespace default:
+./clear-route-status.sh default example
+
+NOTE: if a router that admits a route is running it will immediatly update the cleared route status 
+"
+
+
+if [[ ${#} -ne 2 ]] || [[ ${@} == *" help "* ]]; then
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Quote expansions

```sh
if [[ ${#} -ne 2 ]] || [[ "${@}" == *" help "* ]]; then
                          ^    ^
```

Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: One of the benefits of `[[` over `[` is you can inline boolean logic:

```sh
if [[ ${#} -ne 2 || "${@}" == *" help "* ]]; then
```

File: images/router/clear-route-status.sh [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    printf "$USAGE"
+    exit
+fi
+
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Somewhere here declare your inputs:

```sh
namespace="${1}"
route_name="${2:-}"
```

So you're not using `$1` and `$2` in the rest of the script, which helps for readability.

File: images/router/clear-route-status.sh (hunk 10 above)
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Declare function inputs at the top:

```sh
function clear_status() {
    local namespace="$1"
    local route_name="$2"
    ...
```

File: images/router/clear-route-status.sh [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+oc proxy > /dev/null &
+PROXY_PID=$!
+
+
+if [[ ${2} == "ALL" ]]; then
+    ROUTES=$(oc get routes -n $1 -o name)
+    if [[ -n ${ROUTES} ]]; then
+        for ROUTE in $ROUTES; do
+            clear_status $1 ${ROUTE:6}
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Ah, since you need the kind and name separate for the `curl` call, you can get at the route names themselves with:

```sh
$ oc get routes -o jsonpath='{range .items[*]}{.metadata.name} {end}'
```

Instead of using `-o name`.

File: images/router/clear-route-status.sh [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+        done
+    else
+        echo "No routes found for namespace $1"
+    fi
+else
+    sleep 1
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Why is this necessary?

File: images/router/clear-route-status.sh [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+    printf "$USAGE"
+    exit
+fi
+
+oc proxy > /dev/null &
+PROXY_PID=$!
+
+## attempt to access the proxy until it is online
+until curl -s -X GET http://localhost:8001/oapi/v1/ >/dev/null; do
+    :
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Might want to sleep a little here to rate-limit how fast this churns

File: images/router/clear-route-status.sh [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+done
+
+
+if [[ ${route_name} == "ALL" ]]; then
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Quote all of your expansions: `"${var}"`

File: images/router/clear-route-status.sh [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+    ROUTES=$(oc get routes -o jsonpath='{range .items[*]}{.metadata.name} {end}')
+    if [[ -n ${ROUTES} ]]; then
+        for ROUTE in $ROUTES; do
+            clear_status $1 ${ROUTE}
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: `s/$1/${namespace}/g`

File: images/router/clear-route-status.sh [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+
+# This allows for the clearing of route statuses, routers don't clear the routes status so some may be stale.
+# Upon deletion of the routes status active routers will immediatly update with a vaild status
+
+function clear_status() {
+    local namespace="$1"
+    local route_name="$2"
+    MY_JSON_BLOB=$(oc get --raw http://localhost:8001/oapi/v1/namespaces/${namespace}/routes/${route_name}/)
+    MODIFIED_JSON=$(echo "$MY_JSON_BLOB" | jq 'del(.status.ingress)')
+    curl -s -X PUT http://localhost:8001/oapi/v1/namespaces/$1/routes/$2/status --data-binary "${MODIFIED_JSON}" -H "Content-Type: application/json" > /dev/null
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Don't use the positional args, and make sure the `$namespace` here is not shadowing it from outside -- may not even need to pass it through since there will only ever be one in play.

**Comment by knobunc** [comment]
@stevekuznetsov can you re-review this please

File: images/router/clear-route-status.sh [hunk 21]
```diff
@@ -0,0 +1,117 @@
...
+set -o nounset
+
+# This allows for the clearing of route statuses, routers don't clear the routes status so some may be stale.
+# Upon deletion of the routes status active routers will immediatly update with a vaild status
+
+#clears status of all routers 
+function clear_status() {
+    local namespace="$1"
+    local route_name="$2"
+    MY_JSON_BLOB=$(oc get --raw http://localhost:8001/oapi/v1/namespaces/${namespace}/routes/${route_name}/)
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Good practice not to let anything leak out of function scope, so

```sh
local my_json_blob; my_json_blob="$( ... )"
```

File: images/router/clear-route-status.sh [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+    MODIFIED_JSON=$(echo "$MY_JSON_BLOB" | jq 'del(.status.ingress)')
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: This could use a `local` as well

File: images/router/clear-route-status.sh [hunk 23]
```diff
@@ -0,0 +1,117 @@
...
+    curl -s -X PUT http://localhost:8001/oapi/v1/namespaces/${namespace}/routes/${route_name}/status --data-binary "${MODIFIED_JSON}" -H "Content-Type: application/json" > /dev/null
+    echo "route status for route ${route_name} in namespace ${namespace} cleared"
+}
+
+#sets up clearing a status set by a specific router
+function clear_status_set_by() {
+    local router_name="$1"
+
+    NAMESPACES=$(oc get namespaces -o jsonpath='{range .items[*]}{.metadata.name} {end}')
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Does `-o jsonpath={.items[*].metadata.name}` work here? A little more `jsonpath`-ey and a little less `go-template`-ey

File: images/router/clear-route-status.sh [hunk 24]
```diff
@@ -0,0 +1,117 @@
...
+    for NAMESPACE in $NAMESPACES; do
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: If you're not going to use the `$NAMESPACES` list anywhere else you can just
```sh
for namespace in $( oc get namespaces -o 'jsonpath={.items[*].metadata.name}' ); do
```

File: images/router/clear-route-status.sh [hunk 25]
```diff
@@ -0,0 +1,117 @@
...
+        ROUTES=$(oc get routes -o jsonpath='{range .items[*]}{.metadata.name} {end}' --namespace=$NAMESPACE 2>/dev/null)
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Could be `local`

File: images/router/clear-route-status.sh [hunk 26]
```diff
@@ -0,0 +1,117 @@
...
+        if [[ -n "${ROUTES}" ]]; then
+            for ROUTE in $ROUTES; do
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Best practice to use `$ALL_CAPS` for exported variables and `$lowercase` for local stuff.

```sh
local route
for route in ${routes}; do
```

Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Also better to always have `${varname}` over `$varname`

File: images/router/clear-route-status.sh [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+                clear_routers_status $NAMESPACE $ROUTE $router_name
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Always better to quote and brace all vars

```sh
clear_router_status "${namespace}" "${route}" "${router_name}"
```

File: images/router/clear-route-status.sh [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+    if [[ "${INDEX}" != null ]]; then
+        MODIFIED_JSON=$(echo "$MY_JSON_BLOB" | jq 'del(.status.ingress['${INDEX}'])')
+        curl -s -X PUT http://localhost:8001/oapi/v1/namespaces/${namespace}/routes/${route_name}/status --data-binary "${MODIFIED_JSON}" -H "Content-Type: application/json" > /dev/null
+        echo "route status for route ${route_name} set by router ${router_name} cleared"
+    else
+        echo "route ${route_name} has no status set by ${router_name}"
+    fi
+}
+
+cleanup() {
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: Doesn't follow the same function declaration style as the rest

File: images/router/clear-route-status.sh [hunk 29]
```diff
@@ -0,0 +1,117 @@
...
+    fi
+}
+trap cleanup EXIT
+
+USAGE="Usage:
+To clear only the status set by a specific router on all routes in all namespaces
+./clear-router-status.sh -r [router_name]
+
+router_name is the name in the deployment config, not the name of the pod. If the router is running it will
+immediatly update any cleared status.
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: `s/immediatly/immediately/`

File: images/router/clear-route-status.sh [hunk 30]
```diff
@@ -0,0 +1,117 @@
...
+./clear-route-status.sh defalt ALL
+
+To clear the status of route example in namespace default:
+./clear-route-status.sh default example
+
+NOTE: if a router that admits a route is running it will immediatly update the cleared route status 
+"
+
+if [[ ${#} -ne 2 || "${@}" == *" help "* ]]; then
+    printf "$USAGE"
```
Comment (stevekuznetsov) [inline images/router/clear-route-status.sh]: You're using `${USAGE}` as the format string for `printf` here ... either:

```sh
pritnf "%s" "${USAGE}"
```

or

```sh
echo -n "${USAGE}"
```

File: images/router/clear-route-status.sh [hunk 31]
```diff
@@ -0,0 +1,115 @@
...
+immediately update any cleared status.
+
+To clear the whole status field of the route by route name 
+./clear-route-status.sh [namespace] [route-name | ALL]
+
+To clear the status of all routes in all namespaces:
+oc get namespaces | awk '{if (NR!=1) print \$1}' | xargs -n 1 -I %% ./clear-route-status.sh %% ALL
+
+To clear the status of all routes in namespace default:
+./clear-route-status.sh defalt ALL
```
Comment (knobunc) [inline images/router/clear-route-status.sh]: Typo defalt

**Comment by knobunc** [comment]
@smarterclayton do we need CLI review or anything to add a script to the hack dir?  And is this appropriate to put in hack?

**Comment by knobunc** [comment]
[merge]

**Comment by knobunc** [comment]
Squash please and we can merge

**Comment by knobunc** [comment]
[merge]
